### PR TITLE
S3 uploader fixes

### PR DESCRIPTION
**Pull request by samansmink**
This PR fixes the issue where IOExceptions thrown in an upload thread would not be caught. With this PR, as soon as an IOException is thrown in an upload thread, the exception is stored and rethrown in DuckDB allowing it to be caught and handled.

Additionally, it fixes an issue I ran into where with multiple concurrent connections to duckdb doing uploads, the upload thread limit would be incorrect. I added an extra test case for this.

finally, the current CI had an issue with the scripts/get_test_list.py script that would run both too many tests and not run the _slow tests. This is now made a bit nicer by filtering on the 'require-env S3_TEST_SERVER_AVAILABLE 1' statement, which indicates that it requires the test server to be available.

Note that this approach is still pretty hacky, but should suffice for now until we implement a proper generalized solution of adding IO Thread support in DuckDB which we plan on doing as part of adding Async IO to DuckDB in the (somewhat) near future.

Edit:
Additionally, the CI OOM failure in this PR led me to a bug/unexpected behaviour with the ColumnDataCollection where the Reset function would not actually free the memory. This was a problem with the Parquet Writer which would not free its memory after flushing to disk. The test `upload_large_file.test_slow` now also checks memory usage when writing the files remains reasonable.

### CODE REVIEW

File: test/sql/copy/s3/upload_file_parallel.test_slow [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+    l_shipdate >= CAST('1994-01-01' AS date)
+    AND l_shipdate < CAST('1995-01-01' AS date)
+    AND l_discount BETWEEN 0.05
+    AND 0.07
+    AND l_quantity < 24;
+----
+123141078.2283
+
+# We do this in parallel to also test synchronization of s3fs between 2 connections
+#concurrentloop threadid 0 4
```
Comment (Mytherin) [inline test/sql/copy/s3/upload_file_parallel.test_slow:34]: Should this be commented out?

**Review by Mytherin — COMMENTED** [review]
Thanks for the PR! LGTM. Minor comment below:

File: src/common/types/column_data_collection.cpp [hunk 2]
```diff
@@ -915,6 +915,9 @@ void ColumnDataCollection::Print() const {
 void ColumnDataCollection::Reset() {
 	count = 0;
 	segments.clear();
+
+	// Refreshes the ColumnDataAllocator to prevent holding on to allocated data unnecessarily
+	allocator = make_shared<ColumnDataAllocator>(*allocator);
```
Comment (samansmink) [inline src/common/types/column_data_collection.cpp:920]: @lnkuiper would you mind double checking my fix here for the issue where the ColumnDataCollection  would not properly free data on reset?

Comment (lnkuiper) [reply src/common/types/column_data_collection.cpp:920]: LGTM

**Review by lnkuiper — APPROVED** [review]
LGTM

**Pull request merged**